src/Orleans/Threading/CancellationTokenManager.cs

```diff
@@ -0,0 +1,65 @@
+﻿using System;
+using System.Threading;
+using System.Threading.Tasks;
+using Orleans.Runtime;
+
+namespace Orleans.Threading
+{
+    internal static class CancellationTokenManager
+    {
+        public static void WrapCancellationTokens(object[] arguments)
+        {
+            if (arguments == null) return;
+            for (var i = 0; i < arguments.Length; i++)
+            {
+                var argument = arguments[i];
+                if (argument is CancellationToken)
+                {
+                    arguments[i] = WrapCancellationToken((CancellationToken)argument);
+                }
+            }
+        }
+
+        public static CancellationTokenWrapper WrapCancellationToken(CancellationToken ct)
+        {
+            var tokenId = Guid.NewGuid();
+            return new CancellationTokenWrapper(tokenId, ct)
+            {
+                OnSerialization = () =>
+                {
+                    RegisterTokenCallbacks(ct, tokenId);
+                }
+            };
+        }
+
+        public static Task<CancellationToken> GetCancellationToken(Guid tokenId)
+        {
+            return RuntimeClient.Current.InternalGrainFactory.GetGrain<ICancellationTokenHolderGrain>(tokenId).GetCancellationToken();
+        }
+
+        private static void RegisterTokenCallbacks(CancellationToken ct, Guid tokenId)
+        {
+            if (!ct.CanBeCanceled) return;
+            ct.Register(
+               async (f) =>
+               {
+                   await Cancel(tokenId);
+               },
+                new GCObserver(
```

Comment:
GCObserver is essentially an instance that will be collected with `TaskCompletionSource` that issued the `CancellationToken`, so that linked `TaskCompletionSource` on different silo can be freed aswell